python/ray/data/_internal/block_batching/util.py

```diff
@@ -0,0 +1,87 @@
+import logging
+import queue
+import threading
+from typing import Callable, Iterator, TypeVar
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+logger = logging.getLogger(__name__)
+
+
+def _make_async_gen(
+    base_iterator: Iterator[T],
+    fn: Callable[[Iterator[T]], Iterator[U]],
+    num_workers: int = 1,
+) -> Iterator[U]:
+    """Returns a new iterator with elements fetched from the base_iterator
+    in an async fashion using a threadpool.
+
+    Each thread in the threadpool will fetch data from the base_iterator in a
+    thread-safe fashion, and apply the provided computation.
```

Comment:
To make it extra clear where the concurrent/parallel computation will happen.
```suggestion
    thread-safe fashion, and apply the provided computation concurrently
    (and in parallel, for the bits of fn that release the GIL).
```

Reply:
updated